pkg/collector/collector.go

```diff
@@ -0,0 +1,216 @@
+		fallbackEndpoint := fallbackProtocol + "://" + endpoint
+		protocol, addr, err = parseEndpoint(fallbackEndpoint)
+		if err != nil {
+			return "", "", err
+		}
+	}
+	return protocol, addr, err
+}
+
+func parseEndpoint(endpoint string) (string, string, error) {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", "", fmt.Errorf("error while parsing: %w", err)
+	}
+
+	switch u.Scheme {
+	case "tcp":
+		return "tcp", u.Host, nil
+	case "unix":
+		return "unix", u.Path, nil
+
+	case "":
+		return "", "", fmt.Errorf("using %q as %w", endpoint, ErrEndpointDeprecated)
+
+	default:
+		return u.Scheme, "", fmt.Errorf("%q: %w", u.Scheme, ErrProtocolNotSupported)
+	}
+}
+
+func getImageClient(ctx context.Context, socketPath string) (pb.ImageServiceClient, *grpc.ClientConn, error) {
+	addr, dialer, err := GetAddressAndDialer(socketPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithBlock(), grpc.WithInsecure(), grpc.WithContextDialer(dialer))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	imageClient := pb.NewImageServiceClient(conn)
+
+	return imageClient, conn, nil
+}
+
+func logError(err error) {
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func getImageResult(imageRepoTag []string, imageRepoDigest []string) (imageResult string) {
+	if len(imageRepoTag) == 0 {
+		imageResult = imageRepoDigest[0]
+	} else {
+		imageResult = imageRepoTag[0]
+	}
+	return imageResult
+}
+
+func writeListImagesToCollectorCR(clientSet *kubernetes.Clientset, c Client, socketPath string) (err error) {
+	backgroundContext, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	images, err := c.listImages(backgroundContext)
+	logError(err)
+
+	// list of images repo's
+	imagesResults := make([]string, 0, len(images))
+
+	// Get imageResults slice from repoTags or repoDigest
+	for _, image := range images {
+		imagesResults = append(imagesResults, getImageResult(image.RepoTags, image.RepoDigests))
+	}
+
+	imageCollectorResult := eraserv1alpha1.ImageCollectorResult{
+		TypeMeta: v1.TypeMeta{
+			APIVersion: "eraser.sh/v1alpha1",
+			Kind:       "ImageCollectorResult",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: "imagecollectorresult-" + os.Getenv("NODE_NAME"),
+		},
+		Status: eraserv1alpha1.ImageCollectorResultStatus{
+			Node:          os.Getenv("NODE_NAME"),
+			ImagesResults: imagesResults,
+		},
+	}
+
+	body, err := json.Marshal(imageCollectorResult)
+	logError(err)
+
+	// Create imageCollectorResult object
+	_, err = clientSet.RESTClient().Post().
+		AbsPath(apiPath).
+		Name(imageCollectorResult.Name).
+		Resource("imagecollectorresult").
+		Body(body).DoRaw(backgroundContext)
+
+	if err != nil {
+		log.Println("Could not create imagecollectorresult for node ", os.Getenv("NODE_NAME"))
+		return err
+	}
+
+	return nil
+}
+
+func main() {
+	runtimePtr := flag.String("runtime", "docker", "container runtime")
```

Comment:
i think most of the code is reused in other places, can you refactor so common code is not duplicated but reused?

Reply:
applies to other functions like dial, getImageClient, etc